File: C++/75.颜色分类.cpp
/*
 * @lc app=leetcode.cn id=75 lang=cpp
 *
 * [75] 颜色分类
 */

// @lc code=start
class Solution
{
public:
    void sortColors(vector<int> &nums)
    {
        int len = nums.size();
        int p = 0, q = 0;
        for (int i = 0; i < len; i++)
        {
            if (nums[i] == 1)
            {
                swap(nums[i], nums[q]);
                q++;
            }
            else if (nums[i] == 0)
            {
                swap(nums[i], nums[p]);
                if (p < q)
                    swap(nums[i], nums[q]);
                p++;
                q++;
            }
        }
    }
};
// @lc code=end

File: JavaScript/75.颜色分类.js
/*
 * @lc app=leetcode.cn id=75 lang=javascript
 *
 * [75] 颜色分类
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */

var sortColors = function (nums) {
  let left = 0;
  let right = nums.length - 1;
  let cur = 0;

  while (cur <= right) {
    if (nums[cur] === 0) {
      [nums[cur], nums[left]] = [nums[left], nums[cur]];
      left++;
      cur++;
    } else if (nums[cur] === 2) {
      [nums[cur], nums[right]] = [nums[right], nums[cur]];
      right--;
    } else {
      cur++;
    }
  }

  return nums;
};
// @lc code=end
